src/material-experimental/mdc-tooltip/tooltip.ts

```diff
@@ -0,0 +1,116 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  Directive,
+  ElementRef,
+  Inject,
+  NgZone,
+  Optional,
+  ViewContainerRef,
+  ViewEncapsulation,
+} from '@angular/core';
+import {DOCUMENT} from '@angular/common';
+import {Platform} from '@angular/cdk/platform';
+import {AriaDescriber, FocusMonitor} from '@angular/cdk/a11y';
+import {Directionality} from '@angular/cdk/bidi';
+import {ConnectedPosition, Overlay, ScrollDispatcher} from '@angular/cdk/overlay';
+import {
+  MatTooltipDefaultOptions,
+  MAT_TOOLTIP_DEFAULT_OPTIONS,
+  MAT_TOOLTIP_SCROLL_STRATEGY,
+  _MatTooltipBase,
+  _TooltipComponentBase,
+} from '@angular/material/tooltip';
+import {numbers} from '@material/tooltip';
+import {matTooltipAnimations} from './tooltip-animations';
+
+/**
+ * Directive that attaches a material design tooltip to the host element. Animates the showing and
+ * hiding of a tooltip provided position (defaults to below the element).
+ *
+ * https://material.io/design/components/tooltips.html
+ */
+@Directive({
+  selector: '[matTooltip]',
+  exportAs: 'matTooltip',
+  host: {
+    'class': 'mat-mdc-tooltip-trigger'
+  }
+})
+export class MatTooltip extends _MatTooltipBase<TooltipComponent> {
+  protected readonly _tooltipComponent = TooltipComponent;
+  protected readonly _transformOriginSelector = '.mat-mdc-tooltip';
+
+  constructor(
+    overlay: Overlay,
+    elementRef: ElementRef<HTMLElement>,
+    scrollDispatcher: ScrollDispatcher,
+    viewContainerRef: ViewContainerRef,
+    ngZone: NgZone,
+    platform: Platform,
+    ariaDescriber: AriaDescriber,
+    focusMonitor: FocusMonitor,
+    @Inject(MAT_TOOLTIP_SCROLL_STRATEGY) scrollStrategy: any,
+    @Optional() dir: Directionality,
+    @Optional() @Inject(MAT_TOOLTIP_DEFAULT_OPTIONS) defaultOptions: MatTooltipDefaultOptions,
+
+    /** @breaking-change 11.0.0 _document argument to become required. */
+    @Inject(DOCUMENT) _document: any) {
+
+    super(overlay, elementRef, scrollDispatcher, viewContainerRef, ngZone, platform, ariaDescriber,
+      focusMonitor, scrollStrategy, dir, defaultOptions, _document);
+    this._viewportMargin = numbers.MIN_VIEWPORT_TOOLTIP_THRESHOLD;
+  }
+
+  protected _addOffset(position: ConnectedPosition): ConnectedPosition {
+    const offset = numbers.UNBOUNDED_ANCHOR_GAP;
+    const isLtr = !this._dir || this._dir.value == 'ltr';
+
+    if (position.originY === 'top') {
+      position.offsetY = -offset;
+    } else if (position.originY === 'bottom') {
+      position.offsetY = offset;
+    } else if (position.originX === 'start') {
+      position.offsetX = isLtr ? -offset : offset;
+    } else if (position.originX === 'end') {
+      position.offsetX = isLtr ? offset : -offset;
+    }
+
+    return position;
+  }
+}
+
+/**
+ * Internal component that wraps the tooltip's content.
+ * @docs-private
+ */
+@Component({
+  selector: 'mat-tooltip-component',
```

Comment:
Is this supposed to be `MatTooltipComponent` or `MatTooltipContent`? the comment sounds like it should be called content

Reply:
No idea, it's been like this since the tooltip was first introduced. I don't know whether we can rename it now since the component has been exported so it would be a breaking change.

Reply:
Yeah, it should have been something like "container", but most people never see this component anyway so it doesn't matter much